### DISCUSSION TITLE
Berry Exception

### DISCUSSION THREAD

**Discussion by trlafleur** [arendst/Tasmota#14571] [Q&A] [answered]
I have a test program in Berry, from time to time, I get this exception...  it there a way to catch this in Berry??

~~~
BRY: Exception> 'key_error' - SR04
stack traceback:
	<unknown source>: in function `event`
~~~

My test code... (still needs some work...)

~~~
class FOTank : Driver

  var greenhouse_data
  def greenhouse()

#-   275 gal Tank Array -#
     var Tank275 = [0,2,5,9,14,19,25,31,37,44,51,58,65,72,80,87,94,101,108,115,123,130,137,144,151,158,166,173,180,187,194,201,209,216,223,230,236,243,249,254,260,265,269,272,275]

print ("\n")

    # Read needed Sensor data
    import json
    var sensors=json.load(tasmota.read_sensors())
    var d =sensors['SR04']['Distance']
    print("Dist: ", d)
	
    #  Convert from MM to inch and round up
    var d1 = (d / 2.54) + .5                        # 24.4
    d1 = int(d1)
    if (d1 > 44) d1 = 44 end
    if (d1 <= 0) d1 = 0 end
    print("Inch: ",d1)
    
    var d2 = Tank275.item(d1)
    d2 = int(d2)
    print ("Gal: ",d2)
    
    var d3 = (real(d2) / Tank275.item(44)) * 100
    d3 = int (d3)
    print ("Percent:", d3)
    
    self.greenhouse_data = [d, d1, d2, d3]
    return self.greenhouse_data
  end



  #- *************************************** -#
  def every_second()
	if !self.greenhouse return nil end
	self.greenhouse()
  end
  
   #- *************************************** -#
  def web_sensor()
    import string
    if !self.greenhouse_data return nil end  #- exit if not initialized -#	
    var msg = string.format(
              "{s}Distance{m}%.f mm{e}"..
              "{s}Distance Inch{m}%.f inch{e}"..
              "{s}D2{m}%.f gal{e}"..
              "{s}D3{m}%.2f %%{e}",
              self.greenhouse_data[0],self.greenhouse_data[1],self.greenhouse_data[2],self.greenhouse_data[3])
    tasmota.web_send_decimal(msg)
  end
  
   #- *************************************** -#
  def json_append()
    if !self.greenhouse_data return nil end
	import string
	var msg = string.format(",\"GhData\":{\"SeaPressure2\":%.1f,\"AbsoluteHumidity\":%.2f,\"VPD\":%.2f,\"Humidity Deficit\":%.2f}",
              self.greenhouse_data[0],self.greenhouse_data[1],self.greenhouse_data[2],self.greenhouse_data[3])
    tasmota.response_append(msg)
  end
end

FOTank = FOTank()
tasmota.add_driver(FOTank)
~~~

**Comment by s-hadinger** [comment]
You can use `sensors.find("SR04", {})` to avoid an exception if the key is not present

**Comment by trlafleur** [comment]
thanks... 
oh....  thought this might fix it, but no...
~~~
     # Read needed Sensor data
    import json
    var sensors=json.load(tasmota.read_sensors())
    if (!sensors.find("SR04", {}) ) return nil end
    var d = sensors['SR04']['Distance']
~~~

**Reply by sfromis** [reply]
If you just reentered the driver class, *without* recreating the instance, you'll still be running the old code. If you recreated the instance, and also added the new instance as a driver, you'll still be running the old code, as the old unchanged version will still linger. Simplest fix is to restart Tasmota, to have a clean slate.

**Reply by trlafleur** [reply]
I always reboot the Tasmota prior to reloading the new Berry code... 

An unrelated question, If I load an ESP32 with Tasmota, then if I want to use the same device under Arduino IDE, the code loads, but the ESP32 crashed when it run.... is there a work-around for this ??

**Reply by barbudor** [reply]
Why would you do that 🤔😱
😂
Kidding aside, I would suggest a full erase with esptool while changing from one environment to the other (valid in both direction)

**Comment by trlafleur** [comment]
These Exceptions are still happening, still unable to catch them in my code... The code keeps running, but I would nice to fix this...

Any ideas??

~~~
BRY: Exception> 'key_error' - SR04
stack traceback:
	<unknown source>: in function `event`
~~~


~~~

#- This is using an SR04 ultrasonic sensor to check the current volume
   of a fuel-oil heating tank. The SR04 measures the distance from the
   sensor to the fuel oil and returns a value in centimeters. 
   We then convert this to inches, do a table lookup from the tank 
   manufacturer's datasheet to get the current volume remaining in the tank. 
   We display the data on the web page and send it via MQTT.

   When we start this project we were doing a liner interpolation of the 
   manufacturers table to get a more precise volume, but common sense 
   step in when we realize we just need to know if the tank was full or almost 
   empty, so we now just use a simple table lookup..…

   22Jan2022 TRL 1.0
   tom@lafleur.us
-#


#- *************************************** -#
class FOTank : Driver

   var tank_data
 
  
  
#- *************************************** -#
   def tank()
   
    var tank_offset = -18.0                              # offset from SR04 to top of tank (dead zone...)

# 275 gal Tank Array 45 entries from tank manufacturers table
    var Tank275 = [0,2,5,9,14,19,25,31,37,44,51,58,65,72,80,87,94,101,108,115,123,130,137,144,151,158,166,173,180,187,194,201,209,216,223,230,236,243,249,254,260,265,269,272,275]

    print ("\n")

    # Read Sensor data
    import json
    var sensors = json.load(tasmota.read_sensors())
    
    if (sensors.find("SR04", {}) == nil) return nil end
    var d = sensors['SR04']['Distance']
    print("Dist: ", d)
    d = d + tank_offset
	
    #  Convert from MM/CM to inch and round up
    var d1 = (d / 2.54) + .5                        # 25.4
    d1 = int(d1)                                    # we want an integer here for table lookup
    if (d1 > 44) d1 = 44 end                        # do a bounds check
    if (d1 <= 0) d1 = 0 end
    print("Inch: ",d1)
    
    var d2 = Tank275.item(d1)                       # do table lookup
    d2 = int(d2)
    print ("Gal: ",d2)
    
    var d3 = (real(d2) / Tank275.item(44)) * 100    # calculate percent of full
    d3 = int (d3)
    print ("Percent:", d3)
    
    self.tank_data = [int (d), d1, d2, d3]          # return the data
    return self.tank_data
  end


#- *************************************** -#
  def every_second()
	if !self.tank return nil end
	self.tank()
  end
  
   
#- *************************************** -#
  def web_sensor()
    import string
    if !self.tank_data return nil end               #- exit if not initialized -#	
    var msg = string.format(
              "{s}Distance{m}%.f cm{e}"..
              "{s}Distance{m}%.f inch{e}"..
              "{s}Volume{m}%.f gal{e}"..
              "{s}Percent Full{m}%.f %%{e}",
              self.tank_data[0],self.tank_data[1],self.tank_data[2],self.tank_data[3])
    tasmota.web_send_decimal(msg)
  end
  

#- *************************************** -#
  def json_append()
    if !self.tank_data return nil end
	import string
	var msg = string.format(",\"Tank\":{\"Distance cm\":%.f,\"Distance Inch\":%.f,\"Volume\":%.f,\"Percent Full\":%.f}",
              self.tank_data[0],self.tank_data[1],self.tank_data[2],self.tank_data[3])
    tasmota.response_append(msg)
  end
  
end


#- *************************************** -#
FOTank = FOTank()
tasmota.add_driver(FOTank)

~~~

**Reply by sfromis** [reply]
Try a bit of basic debugging, like doing a `print(sensors)`, to see what you get. You may be getting a `nil` value in the map from the `json.load`, and then it would also make sense to print the output string of your `tasmota.read_sensors()`

Those SR04 are not great sensors, and there may be a risk of Tasmota failing to get a value from it.

**Comment by trlafleur** [comment]
well, it looks like the SR04 key is missing....

I expected that the if (sensors.find("SR04", {}) == nil)  return nil end, to catch this... looks like this may not be doing what I expect...

~~~
 # Read Sensor data
    import json
    var sensors = json.load(tasmota.read_sensors())
    print(sensors)
    if (sensors.find("SR04", {}) == nil)  return nil end
    var d = sensors['SR04']['Distance']
    print("Dist: ", d)
    d = d + tank_offset
~~~

~~~
Good:
{'SR04': {'Distance': 79.316}, 'Time': '2022-01-23T16:39:20', 'TempUnit': 'C', 'ESP32': {'Temperature': 53.3}, 'Tank': {'Percent Full': 54, 'Distance Inches': 24, 'Distance cm': 60, 'Volume': 151}}

Bad:
{'Time': '2022-01-23T16:40:09', 'Tank': {'Percent Full': 54, 'Distance Inches': 24, 'Distance cm': 60, 'Volume': 151}, 'ESP32': {'Temperature': 46.7}, 'TempUnit': 'C'}
BRY: Exception> 'key_error' - SR04
stack traceback:
	<unknown source>: in function `event`
~~~

**Reply by trlafleur** [reply]
By the way, I will most likely not be using the SR04 for this but will use a VL53L1 TOF sensor, waiting for parts...
but, most of this code will be the same...

**Reply by sfromis** [reply]
Yeah, the VL53L1X is a better sensor :-) 

Looks like you'd have gotten the expected result with the simpler 
`if (sensors.find("SR04") == nil)  return nil end`
This is due to `sensors.find("SR04", {})` producing an empty map/list instead of `nil`, when no hit. That is *literally* what you're asking for by having the default value being an empty map/list.

You should have explored the meaning of the suggestion of s-hadinger instead of blindly assuming it to produce `nil` :-D

**Comment by trlafleur** [comment]
I’m new to berry, my C habits are shinning through…

Thanks

On Sun, Jan 23, 2022 at 8:25 AM sfromis ***@***.***> wrote:

> Yeah, the VL53L1X is a better sensor :-)
>
> Looks like you'd have gotten the expected result with the simpler
> if (sensors.find("SR04") == nil) return nil end
> This is due to sensors.find("SR04", {}) producing an empty map instead of
> nil, when no hit.
>
> You should have explored the meaning of the suggestion of s-hadinger
> instead of blindly assuming it to to produce nil :-D
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/arendst/Tasmota/discussions/14571#discussioncomment-2028022>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABC4EK72Z7QCLAR3SJ6OJWLUXQTXLANCNFSM5MSH5RIQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 

~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Reply by trlafleur** [reply]
In the documentation, it says:

"map.find(key)
Get the value mapped by the key. It will return nil when the key-value pair does not exist."

So I would expect:  if !sensors.find("SR04", {})  return end  or  if (sensors.find("SR04", {})  == nil)  return end
 -->  to fail and return

so I'm at a loss on how this should work??
~~~
 # Read Sensor data
    import json
    var sensors = json.load(tasmota.read_sensors())
    #print(sensors)
    if !sensors.find("SR04", {})  return end
    var d = sensors['SR04']['Distance']
    print("Dist: ", d)
~~~

**Reply by sfromis** [reply]
That variation works due to an empty list/map being evaluated as `false` when treated as a boolean. You asked for an empty map/list as default.....

Maybe the docs should've mentioned the feature of specifying a default value, the part you quoted does not match the code you used.

**Comment by trlafleur** [comment]
thanks, I will try that...

~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Comment by trlafleur** [comment]
Ok, I can't seem to catch this error with my limited knowledge of Berry...
I have tried a number of combinations with the if statment..
It's NOT returning nil as documented nor a false..
I'm open to any ideas...
maybe it's a bug??

Thanks

~~~
    count1 = count1 + 1
    print("Count1: ", count1)

    # Read Sensor data
    import json
    var sensors = json.load(tasmota.read_sensors())
    #print(sensors)
    
    #if (sensors.find("SR04", {}) == false)
    #if (sensors.find("SR04", {}) != true)
    #if (!sensors.find("SR04", {})) 
    #if (sensors.find("SR04", {}) == nil) 
    #if !(sensors.find("SR04", {}) == nil)
    #if (sensors.fine("SR04", {}) != nil) 
   
    if (sensors.find("SR04", {}) ) 
        print("OK") 
        #return
    else
        print("Exit5") 
        #return 
    end
    
    count2 = count2 + 1
    print("Count2: ", count2)
~~~

**Reply by barbudor** [reply]
Tom why are you still giving a default value?
`{ }` is neither nil, true nor false, it is an existing but empty object
Just remove it and `find("SR04")` will return nil when not found

```
import json
msg=json.load('{"hello":"world"}')
if (!msg.find("SR04")) print("ko") else print("ok") end
msg=json.load('{"hello":"world","SR04":"found it"}')
if (!msg.find("SR04")) print("ko") else print("ok") end

ko
ok
```

**Reply by s-hadinger** [reply]
I'm sorry for the confusion. `map.find(<key>, <default_value>)` only returns the second <default_value> if the key is not found.

In you former case you did `sensors['SR04']['Distance']`. If using only `sensors.find('SR04')` it would return nil if the key does not exist and the rest would evaluate to `nil['Distance']` which would raise an error.

If you want an if-less version, you can use `sensors.find('SR04', {}).find('Distance')`. This version never raises an exception and returns `nil` if some keys are missing.

Now if you want to check for the existence of a key, just do `sensors.contains('SR04')` which returns either `true` or `fasle`

**Comment by trlafleur** [comment, marked as answer]
Well, it looks like I went deep down a rabbit hole for using a suggestion of `sensors.find("SR04", {})` without first understanding the syntax of `.find` (or understanding Berry)... all is well now, Thank you for your help....

**Comment by btsimonh** [comment]
I'd like to re-open the rabbit hole...  but focus more on the exception catching...

I have the below code.

I'd like a way to remove the driver before re-adding it, if it already existed.
If the driver HAS been added, then `if d1` does not throw, and the driver is successfully removed.
However, if it has not been added, then berry throws with 'syntax_error', and the try/except does not catch it...
I also tried 'except ...'

Any advice?  Does berry exception handling work in Tas? 

Edit: refs:
[try-except docs](https://berry.readthedocs.io/en/latest/source/en/Chapter-4.html?highlight=except#catch-exceptions)
scroll up for an example....

```
# simple UI for calling tas commands.
# note that when pasting into the berry console, you may want to restart between runs..
# since the driver seems to persist multiple times.

import webserver # import webserver class

class MyButtonMethods

  #- create a method for adding a button to the main menu -#
  def web_add_main_button()
    webserver.content_send("<p></p><button onclick='la(\"&m_feed=1\");'>Feed 1</button>")
    webserver.content_send("<p></p><button onclick='la(\"&m_feed=4\");'>Feed 4</button>")
  end

  #- As we can add only one sensor method we will have to combine them besides all other sensor readings in one method -#
  def web_sensor()

    if webserver.has_arg("m_feed")
      print("ui button pressed")
      var val = webserver.arg("m_feed");
      var cmd = "feed "+val;
      tasmota.cmd(cmd);
    end

  end
end

try
  # if this is second run, remove the existing driver.
  if d1
    print("removing existing driver")
    tasmota.remove_driver(d1)
  end
except 'syntax_error'
  # do nothing - normal first run
  print("first run, no driver to remove")
end
  
d1 = MyButtonMethods()
tasmota.add_driver(d1)
print("driver added")
```

**Reply by btsimonh** [reply]
more investigation does not highlight how to work with exceptions.  But it does give a workaround.

If you use global.d1, then there is no exception if d1 does not exist, and it returns the value of d1 if it does exist....

```
# if this is second run, remove the existing driver.
if global.d1
  print("removing existing driver")
  tasmota.remove_driver(global.d1)
else
  # do nothing - normal first run
  print("first run, no driver to remove")
end
  
d1 = MyButtonMethods()
tasmota.add_driver(d1)
print("driver added")
```

**Reply by sfromis** [reply]
Yeah, try/except does not apply to compile time errors (which cannot be trapped), only to errors depending on code execution. 

BTW. You do not need that `: Driver`, that was only in the beginning

**Reply by btsimonh** [reply]
@sfromis - thankyou for confirming.
I'm finding there IS information on Berry out there - it's just really hard to find.
(for other who find this thread) this [discussion](https://github.com/arendst/Tasmota/discussions/17490#discussioncomment-5914758) is useful, and references this [repo](https://github.com/tasmota/Berry_playground), where I've (probably incorrectly) raised an issue containing my simple examples - hopefully they can go into the repo when it's formally started.

**Reply by btsimonh** [reply]
removed ": Driver" here and in the issue to avoid reinforcing it :)

**Reply by s-hadinger** [reply]
I'm not sure what you are trying to do here. Berry checks that a global exists before you can use it. It can be a problem if you know in advance that a global will exist at runtime but does not at compile time.

To deal with this, just use `global.d1` as mentioned above. This bypasses the compile time check

**Reply by sfromis** [reply]
A more general strategy to catch errors from possibly unknown variables (or any other compile time error) is variations of using the compile function.
```
var bad='qwerqwer'
var good='bad'
print('good', compile('return ' + good)())
try
  print('bad', compile('return ' + bad)())
except .. as e,m
  print('compile failure:',e,m)
end
```

**Reply by btsimonh** [reply]
@sfromis - so I think you are saying that an exception based solution to the above would be like:

```
# if this is second run, remove the existing driver.
try
  var fn = compile("tasmota.remove_driver(d1)");
  print("removing existing driver")
  fn()
except .. as e,m
  print('existing driver as d1 did not exist:',e,m)
end
```

where compile would succeed if the global d1 existed, and throw if the global d1 did not yet exist?
(tested and working!!! - yes the global.d1 solution is nicer, but it's good to understandstuff like this which is hard to find examples of...)

**Reply by sfromis** [reply]
To me, the primary way of gaining an understanding of language capabilities is not examples but docs. Also a good place to find "surprising goodies", instead of "waiting" for other code....

**Reply by btsimonh** [reply]
Fully agree...  visited the docs, interpreted what you said, it works.  But not an ideal solutions, global.xxx is better.  
Yet I don't think I would have got there without your pointer...  I would never have visited the compile() fn in berry unless developing a Berry integration! (and @s-hadinger warns people should not go there, I probably agree).
So examples (good or bad) are, for me, key to highlighting what needs to be learned from the docs :).
(have since looked at the berry integration in Tas; some undertaking, and a LOT more to understand before one could think of contributing).
